pkg/kube/namespaces.go

```diff
@@ -2,182 +2,97 @@ package kube
 
 import (
 	"context"
-	"fmt"
-	"time"
 
-	"github.com/juju/errors"
 	"github.com/rancher/prometheus-auth/pkg/data"
+	"github.com/rancher/steve/pkg/accesscontrol"
+	v1 "github.com/rancher/types/apis/core/v1"
+	corev1 "github.com/rancher/wrangler-api/pkg/generated/controllers/core/v1"
+
+	"github.com/juju/errors"
 	log "github.com/sirupsen/logrus"
-	authorization "k8s.io/api/authorization/v1"
-	core "k8s.io/api/core/v1"
-	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/apimachinery/pkg/runtime"
+	k8scorev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/labels"
 	"k8s.io/apimachinery/pkg/util/cache"
-	"k8s.io/apimachinery/pkg/watch"
-	"k8s.io/client-go/kubernetes"
-	clientAuthorization "k8s.io/client-go/kubernetes/typed/authorization/v1"
-	clientCache "k8s.io/client-go/tools/cache"
+	"k8s.io/apiserver/pkg/authentication/user"
 )
 
 const (
-	byTokenIndex     = "byToken"
-	byProjectIDIndex = "byProjectID"
+	ByTokenIndex      = "byToken"
+	ByProjectIDIndex  = "byProjectID"
+	annServiceAccount = "kubernetes.io/service-account.name"
 )
 
 type Namespaces interface {
-	Query(token string) data.Set
+	QueryByUser(info *user.DefaultInfo) data.Set
 }
 
 type namespaces struct {
-	subjectAccessReviewsClient clientAuthorization.SubjectAccessReviewInterface
-	reviewResultTTLCache       *cache.LRUExpireCache
-	secretIndexer              clientCache.Indexer
-	namespaceIndexer           clientCache.Indexer
+	ctx                  context.Context
+	monitoringNamespace  string
+	namespaceCache       corev1.NamespaceCache
+	secrets              *Secrets
+	accessStore          accesscontrol.AccessSetLookup
+	reviewResultTTLCache *cache.LRUExpireCache
+}
+
+func NewNamespaces(ctx context.Context,
+	namespaceCache corev1.NamespaceCache, secrets *Secrets,
+	acl *accesscontrol.AccessStore, monitoringNs string) Namespaces {
+	return &namespaces{
+		ctx:                  ctx,
+		monitoringNamespace:  monitoringNs,
+		namespaceCache:       namespaceCache,
+		secrets:              secrets,
+		accessStore:          acl,
+		reviewResultTTLCache: cache.NewLRUExpireCache(1024),
```

Comment:
Why we need `reviewResultTTLCache` here? I can't find any reference of this field.

Reply:
updated